File: cpp/leetcode/1074. number-of-submatrices-that-sum-to-target.cpp
#include "public.h"

//8976ms, 26.51%
//m行n列, O(m*n*n)
//记录每一行的前i列的和
//然后对列进行双循环, 使用unordered_map, 记录前i行, 某一段列的和值, 如果当前前j行的和值sum - target可以找到或者是0, 说明存在i行至j行间, 某一段列, 矩阵和值为target

class Solution {
public:
	int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
		int res = 0;

		int rows = matrix.size();
		if (rows == 0) return 0;
		int cols = matrix[0].size();

		//计算前?列的和
		vector<vector<int>> colsum(rows, vector<int>(cols + 1, 0)); //注意colsum[0] = 0, dumpvalue
		for (int row = 0; row < rows; ++row)
		{
			for (int col = 0; col < cols; ++col)
			{
				colsum[row][col + 1] = colsum[row][col] + matrix[row][col];
			}
		}

		//主循环, O(m*n*n)
		for (int colstart = 0; colstart < cols; ++colstart)
		{
			for (int colend = colstart; colend < cols; ++colend)
			{
				unordered_map<int, int> history_rowsum;
				history_rowsum[0] = 1;
				int prerowsum = 0;
				for (int row = 0; row < rows; ++row)
				{
					prerowsum += colsum[row][colend + 1] - colsum[row][colstart];
					if (history_rowsum.find(prerowsum - target) != history_rowsum.end())
					{
						res += history_rowsum[prerowsum - target];
					}
					history_rowsum[prerowsum]++;
				}
			}
		}

		return res;
	}
};

int main()
{
	Solution* s = new Solution();
	vector<vector<int>> matrix = { {1,-1},{-1,1} };
	auto res = s->numSubmatrixSumTarget(matrix, 0);
	return 0;
}

File: javascript/1074. number-of-submatrices-that-sum-to-target.js

/*
不推荐, 暴力O(n^2*m^2)
9536ms, 11.11%
*/

/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {number}
 */
var numSubmatrixSumTarget = function(matrix, target) {
    var res = 0;
    
    var sumCol = new Array(matrix.length);
    for (var row = 0; row < matrix.length; ++row) {
        sumCol[row] = new Array();
		sumCol[row][0] = 0;
        for (var col = 0; col < matrix[0].length; ++col)
        {
            sumCol[row][col + 1] = sumCol[row][col] + matrix[row][col];
        }
    }	
	
    for(let leftCol = 0; leftCol < matrix[0].length; leftCol++)
    {
        for (let rightCol = leftCol; rightCol < matrix[0].length; rightCol++)
        {
            for (let upRow = 0; upRow < matrix.length; upRow++)
            {
				var presum = sumCol[upRow][rightCol + 1] - sumCol[upRow][leftCol];
                if (presum == target)
                    res++;
                for (let downRow = upRow + 1; downRow < matrix.length; downRow++)
                {
					presum += sumCol[downRow][rightCol + 1] - sumCol[downRow][leftCol];
                if (presum == target)
                    res++;
                }
            }
        }
    }
    return res;
};